Add torch.library.impl_abstract

zou3519 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #110011
* __->__ #109912

Changelog:
- torch.library.impl_abstract optionally accepts a torch.library.Library
  object. If passed in, then the lifetime of the registration is tied to
  the Library object.
- we've also changed torch.library.impl_abstract to work on all
  operators, including overloads.
- we refactored the `torch._custom_ops.*` and `torch._custom_op.*`
  impl_abstract APIs and put them under torch._library. This is the
  final resting place for them. I will follow-up with deleting
  all the `torch._custom_ops.*` stuff later.
- There is a new "SimpleOperatorRegistry" where we actually collect the
  abstract_impl. We will expand this to also hold the other
  torch._custom_ops.* APIs when we move those to torch.library

NB: Previously we had designed
`impl_abstract` assuming a very high-level Python-only custom op API.
We've revisited that since; now, impl_abstract works for all custom ops,
no matter python or C++, no matter the schema. The new refactored design
reflects this better.

Test Plan:
- existing and new tests

zou3519 committed:
Add torch.library.impl_abstract

Changelog:
- torch.library.impl_abstract optionally accepts a torch.library.Library
  object. If passed in, then the lifetime of the registration is tied to
  the Library object.
- we've also changed torch.library.impl_abstract to work on all
  operators, including overloads.
- we refactored the `torch._custom_ops.*` and `torch._custom_op.*`
  impl_abstract APIs and put them under torch._library. This is the
  final resting place for them. I will follow-up with deleting
  all the `torch._custom_ops.*` stuff later.
- There is a new "SimpleOperatorRegistry" where we actually collect the
  abstract_impl. We will expand this to also hold the other
  torch._custom_ops.* APIs when we move those to torch.library

NB: Previously we had designed
`impl_abstract` assuming a very high-level Python-only custom op API.
We've revisited that since; now, impl_abstract works for all custom ops,
no matter python or C++, no matter the schema. The new refactored design
reflects this better.

Test Plan:
- existing and new tests

[ghstack-poisoned]

zou3519 commented on torch/_library/abstract_impl.py:
```diff
@@ -0,0 +1,197 @@
+                f"data-dependent sizes of >= 2"
+            )
+
+        result = self._shape_env.create_unbacked_symint()
+        torch.fx.experimental.symbolic_shapes.constrain_range(result, min=2, max=max)
```
@ezyang I never really understood, but, what happens if someone tries to torch.compile(dynamic=True) a call to custom::nonzero that actually returns a Tensor of size (1, 2) (where the symint represents the 1)?

Does this result in a runtime assertion, and if so, where? (inductor?)

ezyang replied:
So, we're fixing this right now in https://github.com/pytorch/pytorch/pull/108372 but yes, as of today, if runtime assertions were working (they only work in export right now) you'd end up with a `>= 2` assertion in the graph which is wrong.

The fixed version of this is that we compile time assume >= 2, but the runtime assert we generate is >= 0 and all is well. This is for the constrain_as_size combinator, which we should be encouraging people to use. Actually, with new improvements to implicit size assumptions, this isn't even necessary if you send the unbacked symint straight to a constructor.

ezyang replied:
Yeah, you should use `constrain_as_size` here

zou3519 replied:
That is a WIP PR, so I think I am going to land https://github.com/pytorch/pytorch/pull/109912 first, and then we change it to use `sym_constrain_range_for_size` in that PR

zou3519 commented:
NB: this PR is kind of long. I'm happy to attempt to split it up if necessary. It might be easier to just review all of the green lines.

ezyang commented:
I will... eventually review it... it will just be faster if it's shorter 📦

zou3519 commented:
Okay let me split

ezyang commented on torch/_library/utils.py:
```diff
@@ -0,0 +1,38 @@
+
+    def __init__(self, func: Callable, source: str):
+        self.func: Callable = func
+        self.source: str = source
```
nit: make this a dataclass?

ezyang commented on torch/_custom_op/impl.py:
```diff
@@ -779,126 +780,6 @@ def compare(sig_args, schema_args):
-        self._shape_env = _shape_env
-        self._op = _op
-
-    def create_unbacked_symint(self, *, min=2, max=None) -> torch.SymInt:
```
Moved and rewritten at torch/library.py

ezyang commented on torch/_library/abstract_impl.py:
```diff
@@ -0,0 +1,197 @@
+                f"already has an DispatchKey::Meta implementation via a "
+                f"pre-existing torch.library or TORCH_LIBRARY registration. "
+                f"Please either remove that registration or don't call "
+                f"impl_abstract."
```
Hmm, you might legitimately want to override an incomplete meta with a fake tensor impl that can handle data dependent dynamic shapes. Well, we can deal with it if it ever happens.

ezyang commented on torch/_library/abstract_impl.py:
```diff
@@ -0,0 +1,197 @@
+                f"such meta implementation and this error is the correct "
+                f"behavior. Otherwise, please remove the call to get_ctx() "
+                f"in the implementation registered with impl_abstract "
+                f"at {source}"
```
huh, is this a thing lol

zou3519 replied:
Meta formulas can't use the get_ctx() because there isn't a shape env. There are a couple of things we can do here:
- get_ctx() for meta errors out (what is here)
- get_ctx() for meta returns None
- get_ctx().new_dynamic_size() errors.

and we went with the first one

ezyang replied:
I guess my point is, I would only do get_ctx() if I was about to do data dependent output. So the case where I can just "remove get_ctx()" seems unlikely, and not worth calling out

zou3519 replied:
Fair enough

ezyang commented on torch/_library/abstract_impl.py:
```diff
@@ -0,0 +1,197 @@
+                f"ctx.create_dynamic_size(min={min}, ...): expected min to be "
+                f"greater than or equal to 2. PyTorch only supports new "
+                f"data-dependent sizes of >= 2"
+            )
```
I don't think we need this error test anymore, if you say min=1 the idea is that you still compile time >= 2, but runtime is now >= 1 rather than >= 0. Seems legit.

cc @tugsbayasgalan

ezyang replied:
yup

ezyang reviewed (approved):
thx

zou3519 committed:
Update on "Add torch.library.impl_abstract"


Changelog:
- torch.library.impl_abstract optionally accepts a torch.library.Library
  object. If passed in, then the lifetime of the registration is tied to
  the Library object.
- we've also changed torch.library.impl_abstract to work on all
  operators, including overloads.
- we refactored the `torch._custom_ops.*` and `torch._custom_op.*`
  impl_abstract APIs and put them under torch._library. This is the
  final resting place for them. I will follow-up with deleting
  all the `torch._custom_ops.*` stuff later.
- There is a new "SimpleOperatorRegistry" where we actually collect the
  abstract_impl. We will expand this to also hold the other
  torch._custom_ops.* APIs when we move those to torch.library

NB: Previously we had designed
`impl_abstract` assuming a very high-level Python-only custom op API.
We've revisited that since; now, impl_abstract works for all custom ops,
no matter python or C++, no matter the schema. The new refactored design
reflects this better.

Test Plan:
- existing and new tests

[ghstack-poisoned]

zou3519 committed:
Update on "Add torch.library.impl_abstract"


Changelog:
- torch.library.impl_abstract optionally accepts a torch.library.Library
  object. If passed in, then the lifetime of the registration is tied to
  the Library object.
- we've also changed torch.library.impl_abstract to work on all
  operators, including overloads.
- we refactored the `torch._custom_ops.*` and `torch._custom_op.*`
  impl_abstract APIs and put them under torch._library. This is the
  final resting place for them. I will follow-up with deleting
  all the `torch._custom_ops.*` stuff later.
- There is a new "SimpleOperatorRegistry" where we actually collect the
  abstract_impl. We will expand this to also hold the other
  torch._custom_ops.* APIs when we move those to torch.library

NB: Previously we had designed
`impl_abstract` assuming a very high-level Python-only custom op API.
We've revisited that since; now, impl_abstract works for all custom ops,
no matter python or C++, no matter the schema. The new refactored design
reflects this better.

Test Plan:
- existing and new tests

[ghstack-poisoned]

zou3519 commented:
@pytorchbot merge
